frostsnap_coordinator/src/usb_serial_manager.rs

```diff
@@ -536,6 +554,92 @@ impl UsbSerialManager {
     pub fn devices_by_ports(&self) -> &HashMap<String, Vec<DeviceId>> {
         &self.reverse_device_ports
     }
+
+    /// The firmware digest the device has declared it has
+    pub fn firmware_digest_for_device(&self, device_id: DeviceId) -> Option<FirmwareDigest> {
+        self.device_ports
+            .get(&device_id)
+            .map(|device_port| device_port.firmware_digest)
+    }
+
+    pub fn upgrade_bin(&self) -> FirmwareBin {
+        self.firmware_bin
+    }
+
+    pub fn run_firmware_upgrade(&mut self) -> impl Iterator<Item = anyhow::Result<f32>> + '_ {
+        let n_chunks = ceil_div(
+            self.firmware_bin.bin.len() as u32,
+            FIRMWARE_UPGRADE_CHUNK_LEN,
+        );
+        let total_chunks = n_chunks * self.ready.len() as u32;
+
+        let mut iters = vec![];
+
+        for (port_index, (port, io)) in self.ready.iter_mut().enumerate() {
+            let res = io.send_message(&ReceiveSerial::Message(CoordinatorSendMessage {
+                target_destinations: Destination::All,
+                message_body: CoordinatorSendBody::Upgrade(
+                    CoordinatorUpgradeMessage::EnterUpgradeMode,
+                ),
+            }));
+
+            // give some time for devices to forward things and enter upgrade mode
+            std::thread::sleep(Duration::from_millis(100));
+
+            if let Err(e) = res {
+                event!(
+                    Level::ERROR,
+                    port = port,
+                    error = e.to_string(),
+                    "unable to send firmware upgrade initialiazation message"
+                );
+                continue;
+            }
+
+            event!(Level::INFO, port = port, "starting writing firmware");
+            let mut chunks = self
+                .firmware_bin
+                .bin
+                .chunks(FIRMWARE_UPGRADE_CHUNK_LEN as usize)
+                .enumerate();
+
+            iters.push(core::iter::from_fn(move || {
+                let (i, chunk) = chunks.next()?;
+                if let Err(e) = io.raw_write(chunk) {
+                    event!(
+                        Level::ERROR,
+                        port = port,
+                        error = e.to_string(),
+                        "writing firmware failed"
+                    );
+                    return Some(Err(e.into()));
+                }
+                let mut byte = [0u8; 1];
+
+                match io.raw_read(&mut byte[..]) {
+                    Ok(_) => {
+                        if byte[0] != FIRMWARE_NEXT_CHUNK_READY_SIGNAL {
```

Comment:
Do we ever get anything else written upstream other than this signal?

Reply:
not while updating of course.